### PR TITLE
React week2 (Mo Al-Qusi)

### PR DESCRIPTION
**Pull request by alqusi1992**

### CODE REVIEW

File: week2/hackyourweather/src/component/DataProfile.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        <p>Timezone :{timezone}</p>
+        <p>Population:{population}</p>
+        <p>Location:{location}</p>
+        <p>
+          Temp_max:
+          {Object.entries(tmep_max).map((e) => [e[1].main.temp_max])}
+        </p>
+        <p>
+          Temp_min:
+          {Object.entries(tmep_min).map((e) => [e[1].main.temp_min])}
```
Comment (Meno-101) [inline week2/hackyourweather/src/component/DataProfile.js:27]: these two statements (js not the jsx) could have been reduced into two variables and two js use of these variables.

with no magical use of `Object.entries`  and picking only the first entry in the map in which it's already returning an array but we return another array because we are a little tired of thinking of it.

consider the following:
```jsx
function DataProfile({
  ...
  tmep_max,
  tmep_min,
}) {
  // min and max are now your intended objects, use them wisely.
  const {response: max} = tmep_max
  const {response: min} = tmep_min
  return (
    ...
  )
```

replace the dots with the appropriate code and fill the missing parts, **And look more into destructuring** _important_

one last thing, this will work after you fix the previous state problem, told you it will make unexpected behaviour.

File: week2/hackyourweather/src/component/WeatherData.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import DataProfile from "./DataProfile";
+
+function WeatherData() {
+  const [country, setCountry] = useState("");
+  const [stad, setStad] = useState("");
+  const [city, setCity] = useState("");
+  const [timezone, setTimezone] = useState("");
+  const [population, setPopulation] = useState("");
+  const [location, setLocation] = useState("");
+  const [tmep_max, setTmep_max] = useState("");
```
Comment (Meno-101) [inline week2/hackyourweather/src/component/WeatherData.js:12]: tmep? or temp?

naming is important for readability and indexing your code (going through or searching).

File: week2/hackyourweather/src/component/WeatherData.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import React, { useState, useEffect } from "react";
+import axios from "axios";
+import DataProfile from "./DataProfile";
+
+function WeatherData() {
+  const [country, setCountry] = useState("");
+  const [stad, setStad] = useState("");
+  const [city, setCity] = useState("");
```
Comment (Meno-101) [inline week2/hackyourweather/src/component/WeatherData.js:8]: you gotta make a little more accurate names, accurate to their functionality/behaviour/content it's all one harmony.

what does it do? what does it contain? why do I need it?

questions regarding the names, and you will find them.

**Review by Meno-101 — COMMENTED** [review]
Good work,

I have gone your project real quick and left some comments for you to resolve.

I also went through your exercises and I wished to see the same care applied in both, since I saw some good work in your exercises but also some unclear level work in your project I will ask you to update the required from the comments below.

I will approve when I see the changes.

File: week2/hackyourweather/src/component/WeatherData.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  const [timezone, setTimezone] = useState("");
+  const [population, setPopulation] = useState("");
+  const [location, setLocation] = useState("");
+  const [tmep_max, setTmep_max] = useState("");
+  const [tmep_min, setTmep_min] = useState("");
```
[the remark this answers is not in the document]
Comment (alqusi1992) [reply week2/hackyourweather/src/component/WeatherData.js]: You are right, it's now more readable and clean code.

File: week2/hackyourweather/src/component/WeatherData.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  const getWeatherData = (city) => {
+    axios(
+      `https://api.openweathermap.org/data/2.5/forecast?q=${city},&appid=${process.env.REACT_APP_OPENWEATHERMAP_API_KEY}`
+    )
+      .then((response) => {
```
[the remark this answers is not in the document]
Comment (alqusi1992) [reply week2/hackyourweather/src/component/WeatherData.js]: I tried to use async/await, but I get an error, however it seems now an error in my code.

**Comment by alqusi1992** [comment]
Thank you @Meno-101 for reviewing the homework. I have been adjustment all what you have you mentioned in the comment.

**Comment by Meno-101** [comment]
thank you for submitting the changes I will run them today in the evening, and get back to you asap.

**Review by Meno-101 — COMMENTED** [review]
I'm not sure if I can approve, I don't wanna be strict, but you broke part of your program.

I don't wanna argue about the end result of your application, because there are things that are more important now, (you can code better right?)

it's supposed to be a weather app, you made a city population and sunset sundown app, for me as a user it's misleading, but still we have more things to focus on.

File: week2/hackyourweather/src/component/WeatherData.js [hunk 6]
```diff
@@ -1,63 +1,42 @@
...
-        setLocation(response.data.city.sunset);
-        setTmep_max({ response: response.data.list[0] });
-        setTmep_min({ response: response.data.list[0] });
-      })
-      .catch((error) => {
-        console.log(error);
+      .then((res) => res.json())
+      .then((result) => {
+        setWeather(result.city);
       });
```
[the remark this answers is not in the document]
Comment (alqusi1992) [reply week2/hackyourweather/src/component/WeatherData.js]: Yes. You right. I have removed and updated with async, await.

File: week2/hackyourweather/src/component/WeatherData.js [hunk 7]
```diff
@@ -1,63 +1,42 @@
...
   };
 
-  useEffect(() => {
-    getWeatherData();
-  }, []);
+  const submitHandler = (e) => {
+    e.preventDefault();
+    search();
+    setWeather("");
```
[the remark this answers is not in the document]
Comment (alqusi1992) [reply week2/hackyourweather/src/component/WeatherData.js]: I have  fixed the application again. Sorry for that.😅

**Comment by alqusi1992** [comment]
Thank you @Meno-101 for your tips that allow me to type better code.

File: week2/hackyourweather/src/component/WeatherData.js [hunk 8]
```diff
@@ -1,39 +1,41 @@
...
-          <DataProfile weather={weather} />
-          <button className="btn" type="submit">
-            Search
-          </button>
+
+          {Object.entries(weather).length !== 0 ? (
+            <DataProfile weather={weather} />
+          ) : (
+            <h3> Please Enter City Name </h3>
+          )}
```
Comment (Meno-101) [inline week2/hackyourweather/src/component/WeatherData.js:37]: if it doesn't have a length or entries, doesn't mean there is no city or data, it throwed again.

you are not checking your data, have you tried to `console.log` your data "weather"?

luckily this is not a real production project, because if it was and this is in production with the few errors from the past, it's problem.

regardless, we're here to learn, and for the purpose of learning I want you to have another look at your data, I want you to know what do you have and how to deal with it, I will not approve until you get it right 🙂 

in short **inspect your data**, and always check for possible errors, because you will never know how your user is using you application ( users are unpredictable specie, laugh it's good for your health :) )

File: week2/hackyourweather/src/component/DataProfile.js [hunk 9]
```diff
@@ -1,16 +1,23 @@
 import React from "react";
 
 function DataProfile({ weather }) {
+  const cityName = weather.city.name;
+  const countryName = weather.city.country;
+  const minTemp = weather.list[0].main.temp_min;
+  const maxTemp = weather.list[0].main.temp_max;
+  const weatherDesc = weather.list[0].weather[0].description;
+  const weatherMain = weather.list[0].weather[0].main;
```
Comment (Meno-101) [inline week2/hackyourweather/src/component/DataProfile.js:9]: very satisfied to see a weather application over here ❤️

**Review by Meno-101 — COMMENTED** [review]
I'm honestly trying not to be mean, but I want you to get it right, you're not checking on your data, and therefore your'e not knowing how to deal with them, and I want to point out this for you.

because no one will grab your hand when you work and tell you "oh this is what you will receive from the server", sometimes you will never even know, unless you inspect it your self.

File: week2/hackyourweather/src/component/WeatherData.js (hunk 8 above)
Comment (alqusi1992) [reply week2/hackyourweather/src/component/WeatherData.js:37]: If I remove object. Entries then I get an error name is undefined because we are handling with object that why I need to use it. Yes, I use console.log, otherwise I will not see the data on the application.

Comment (Meno-101) [reply week2/hackyourweather/src/component/WeatherData.js:37]: you don't know your conditions! from your answer:

> If I remove object. Entries then I get an error name is undefined

yes, this will definitely throw 👇 
```jsx
          {/*Object.entries(weather).length !== 0 ? (*/
            <DataProfile weather={weather} />
          /* ) : (
            <h3> Please Enter City Name </h3>
          )*/}
```

and no you don't need to use it, look I'm obliged not to give the answer, because if I did I will be the one who is ruining part of your career, as it will take time from your internship (at least) to figure it out.

The only thing I can do is giving you a hint, please don't get me wrong but you're probably thinking in the wrong direction, so I'm gonna give you a final answer because we're already a week late and you need to move on, and I'm gonna try to make it a little more clearer by not giving you the answer, because debugging is your job, most of your time as a developer **you will be debugging** giving you the answer now will never make you a developer **but you will**

----

and please don't get mad at me as I'm speaking in the very basics perhaps regarding your level as for now, because you need it, I wanna cover all aspects and you need to look for what you missed **"yes! you missed something"**.
first off, you only need to see your data, and then act accordingly, this is how it work:
```jsx
          //before the return
         console.log(data) // or weather or whatever
          return /* some jsx */
```

now look at your console, and try to identify what is the key that's gonna be repeated in every request, either was an error of was a success.

now you make your conditions regarding the result you got from the `console.log`, and therefore you will realize you don't need that `Object.entries`.

any more word from me, I will be giving you the answer, I helped as much as I could, it's your turn, and again now you have time to learn to debug.

Comment (alqusi1992) [reply week2/hackyourweather/src/component/WeatherData.js]: NO, I'm not mad of you. I'm learning from my mistake, then I will be better. 
I just removed object. Entries and replace it.

File: week2/hackyourweather/src/component/WeatherData.js [hunk 10]
```diff
@@ -30,10 +30,10 @@ function WeatherData() {
...
           />
 
-          {Object.entries(weather).length !== 0 ? (
-            <DataProfile weather={weather} />
+          {weather.cod === "200" ? (
+            <DataProfile weather={weather} setWeather={setWeather} />
           ) : (
-            <h3> Please Enter City Name </h3>
+            <p> Please Enter City Name</p>
           )}
```
Comment (Meno-101) [inline week2/hackyourweather/src/component/WeatherData.js]: Yay 🎉

File: week2/hackyourweather/src/component/WeatherData.js (hunk 8 above)
Comment (Meno-101) [reply week2/hackyourweather/src/component/WeatherData.js]: Please don't get me wrong, it's the force of the habit, it's just me usually being so strict, and that habit isn't so good, so I'm trynna getting you to do what you need but also not feeling too bad of my self.

thank you though :)

**Review by Meno-101 — APPROVED** [review]
you did it! 
your closest friend is your console, and your best friend is google, never forget that.

File: week2/hackyourweather/src/component/WeatherData.js (hunk 8 above)
Comment (Meno-101) [reply week2/hackyourweather/src/component/WeatherData.js]: Please don't get me wrong, it's the force of the habit, it's just me usually being so strict, and that habit isn't so good, so I'm trynna getting you to do what you need but also not feeling too bad of my self.

thank you though :)

File: week2/hackyourweather/src/component/WeatherData.js (hunk 10 above)
Comment (Meno-101) [reply week2/hackyourweather/src/component/WeatherData.js]: one side note, you didn't use setWeather in your component, (you don't need it, don't pass it.)

no changes required, just a note.

**Comment by alqusi1992** [comment]
Thank you @Meno-101 it takes time to make it, but finally I did it.
